graphbolt/src/cuda/index_select_impl.cu

```diff
@@ -48,6 +53,290 @@ std::pair<torch::Tensor, torch::Tensor> Sort(
+
+  auto input_buffer_it = thrust::make_transform_iterator(
+      iota, IteratorFunc<indptr_t, indices_t>{sliced_indptr, indices});
+  auto output_buffer_it = thrust::make_transform_iterator(
+      iota, IteratorFunc<indptr_t, indices_t>{sub_indptr, sub_indices});
+  auto buffer_sizes = thrust::make_transform_iterator(
+      iota, ConvertToBytes<indptr_t, indices_t>{in_deg});
+  constexpr int64_t max_copy_at_once = std::numeric_limits<int32_t>::max();
+  // Performs the copy from indices into sub_indices.
+  for (int64_t i = 0; i < num_nodes; i += max_copy_at_once) {
+    size_t workspace_size = 0;
+    CUDA_CALL(cub::DeviceMemcpy::Batched(
+        nullptr, workspace_size, input_buffer_it + i, output_buffer_it + i,
+        buffer_sizes + i, std::min(num_nodes - i, max_copy_at_once), stream));
+    auto temp = allocator.AllocateStorage<char>(workspace_size);
+    CUDA_CALL(cub::DeviceMemcpy::Batched(
+        temp.get(), workspace_size, input_buffer_it + i, output_buffer_it + i,
+        buffer_sizes + i, std::min(num_nodes - i, max_copy_at_once), stream));
+  }
+}
+
+std::tuple<torch::Tensor, torch::Tensor> IndexSelectCSCImpl(
+    torch::Tensor indptr, torch::Tensor indices, torch::Tensor nodes) {
+  auto allocator = cuda::BuildAllocator();
+  auto stream = c10::cuda::getDefaultCUDAStream();
+
+  const int64_t num_nodes = nodes.size(0);
+
+  // Output indptr for the slice indexed by nodes.
+  torch::Tensor sub_indptr =
+      torch::empty(num_nodes + 1, nodes.options().dtype(indptr.scalar_type()));
+  torch::Tensor sub_indices;
+  AT_DISPATCH_INTEGRAL_TYPES(
+      indptr.scalar_type(), "IndexSelectCSCIndptr", ([&] {
+        using indptr_t = scalar_t;
+        auto [in_deg_ptr, sliced_indptr_ptr] =
+            ComputeDegree(indptr.data_ptr<indptr_t>(), nodes, stream);
+        auto in_deg = in_deg_ptr.get();
+        auto sliced_indptr = sliced_indptr_ptr.get();
+        {  // Compute the output indptr, sub_indptr.
+          size_t workspace_size = 0;
+          CUDA_CALL(cub::DeviceScan::ExclusiveSum(
+              nullptr, workspace_size, in_deg, sub_indptr.data_ptr<indptr_t>(),
+              num_nodes + 1, stream));
+          auto temp = allocator.AllocateStorage<char>(workspace_size);
+          CUDA_CALL(cub::DeviceScan::ExclusiveSum(
+              temp.get(), workspace_size, in_deg,
+              sub_indptr.data_ptr<indptr_t>(), num_nodes + 1, stream));
+        }
+        // Number of edges being copied
+        indptr_t hop_size;
+        CUDA_CALL(cudaMemcpyAsync(
+            &hop_size, sub_indptr.data_ptr<indptr_t>() + num_nodes,
+            sizeof(hop_size), cudaMemcpyDeviceToHost, stream));
+        // blocking read of hop_size
+        CUDA_CALL(cudaStreamSynchronize(stream));
+        // Allocate output array of size number of copied edges.
+        sub_indices = torch::empty(
+            hop_size, nodes.options().dtype(indices.scalar_type()));
+        GRAPHBOLT_DISPATCH_ELEMENT_SIZES(
```

Comment:
Can we use Pytorch macro to switch the `indices_t`? And I think it is better to switch the index type at the beginning.
```cpp
AT_DISPATCH_INTEGRAL_TYPES(
      indptr.scalar_type(), "IndexSelectCSCIndptr", ([&] {
           AT_DISPATCH_INTEGRAL_TYPES(
           indices.scalar_type(), "IndexSelectCSCIndices", ([&] {
}
}
```

Reply:
Then the code would work for edge probabilities for example.

Reply:
Also, if we dispatch at the beginning, the compile times would be longer and that would be a waste of resources.

Reply:
A good rule of thumb is to contain as little code as possible in the dispatch section.